cmd/junitreport/pkg/builder/nested/test_suites_builder.go

```diff
@@ -0,0 +1,158 @@
+)
+
+// nestedTestSuitesBuilder is a test suites builder that nests suites under a root suite
+type nestedTestSuitesBuilder struct {
+	testSuites *api.TestSuites
+}
+
+// AddSuite adds a test suite to the test suites collection being built and updates the metrics
+// of any parent suites to subsume those of the new suite. If the parent of the test suite to be
+// added is found in the collection, the test suite is added as a child of that suite. Otherwise,
+// parent suites are created by successively removing one layer of package specificity until the
+// root name is found. For instance, if the suite named "root/package/subpackage/subsubpackage"
+// were to be added to an empty collection, the  suites named "root", "root/package", and "root/
+// package/subpackage" would be created and added first, then the suite could be added as a child
+// of the latter parent package. All test suites to be added are asssumed to be nested under one
+// of the root suites created by the constructor method and the attempted addition of a suite not
+// rooted in those suites will fail silently to allow for selective tree-building given a root.
+func (b *nestedTestSuitesBuilder) AddSuite(suite *api.TestSuite) error {
+	parent, err := b.findOrCreateParent(suite.Name)
+	if err != nil {
+		if errors.IsSuiteOutOfBoundsError(err) {
+			// if we were asked to add a new test suite that isn't a child of any current root,
+			// we ignore the request but do not throw an error
+			return nil
+		}
+		return err
+	}
+	parent.Children = append(parent.Children, suite)
+
+	b.updateMetrics(suite)
+
+	return nil
+}
+
+// findOrCreateParent will find or create the parent for the test suite with the given name
+func (b *nestedTestSuitesBuilder) findOrCreateParent(name string) (*api.TestSuite, error) {
+	if !b.isChildOfRoots(name) {
+		return nil, errors.NewSuiteOutOfBoundsError(name)
+	}
+	parentName, exists := b.getParentName(name)
+	if !exists {
+		return nil, fmt.Errorf("attempted to add a new root to the tree: %s", name)
+	}
+
+	if parent := b.findSuite(parentName); parent == nil {
+		// no parent is currently registered, we need to create it
+		parent = &api.TestSuite{
+			Name: parentName,
+		}
+
+		// once we've created the parent package, we need to put it into the tree
+		grandParent, err := b.findOrCreateParent(parentName)
+		if err != nil {
+			return nil, err
+		}
+		grandParent.Children = append(grandParent.Children, parent)
+
+		return parent, nil
+	} else {
+		return parent, nil
+	}
+}
+
+// getParentName returns the name of the parent package, if it exists in the multitree
+func (b *nestedTestSuitesBuilder) getParentName(name string) (string, bool) {
+	if !strings.Contains(name, TestSuiteNameDelimeter) {
+		return "", false
+	}
+
+	delimeterIndex := strings.LastIndex(name, TestSuiteNameDelimeter)
+	return name[0:delimeterIndex], true
+}
+
+func (b *nestedTestSuitesBuilder) isChildOfRoots(name string) bool {
+	for _, rootSuite := range b.testSuites.Suites {
```

Comment:
how does this handle cases where you want to consume everything (aren't specifying `--roots`)?